File: app/src/main/kotlin/com/mobileappdev/homeworkplanner/ClassListFragment.java
package com.mobileappdev.homeworkplanner;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.R;

import java.util.List;

public class ClassListFragment extends Fragment {

    private RecyclerView mCrimeRecyclerView;
    private ClassAdapter mAdapter;
    private Button mAddClassButton;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_class_list, container, false);

        mCrimeRecyclerView = (RecyclerView) view.findViewById(R.id.class_recycler_view);
        mCrimeRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mAddClassButton = (Button) view.findViewById(R.id.add_class_button);
        mAddClassButton.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v){
                Intent intent = new Intent(getActivity(), AddClassActivity.class);
                startActivity(intent);
            }}
        );

        updateUI();

        return view;
    }

    private class ClassHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView mTitleTextView;
        private Class mClass;


        @Override
        public void onClick(View view) {
            Toast.makeText(getActivity(), mClass.getTitile() + " clicked!", Toast.LENGTH_SHORT).show();
        }

        public ClassHolder(LayoutInflater inflater, ViewGroup parent){
            super(inflater.inflate(R.layout.list_item_class, parent, false));
            itemView.setOnClickListener(this);

            mTitleTextView = (TextView)itemView.findViewById(R.id.class_title);
        }

        public void bind(Class aClass){
            mClass = aClass;
            mTitleTextView.setText(mClass.getTitile());
        }
    }

    private void updateUI(){
        ClassSchedule classSchedule = ClassSchedule.get(getActivity());
        List<Class> classes = classSchedule.getClasses();

        mAdapter = new ClassAdapter(classes);
        mCrimeRecyclerView.setAdapter(mAdapter);
    }

    private class ClassAdapter extends RecyclerView.Adapter<ClassHolder>{
        private List<Class> mClasses;

        public ClassAdapter(List<Class> aClasses) { mClasses = aClasses; }

        @NonNull
        @Override
        public ClassHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());

            return new ClassHolder(layoutInflater, parent);
        }

        @Override
        public void onBindViewHolder(@NonNull ClassHolder holder, int position) {
            Class aClass = mClasses.get(position);
            holder.bind(aClass);
        }

        @Override
        public int getItemCount() {
            return mClasses.size();
        }
    }
}

File: app/src/main/kotlin/com/mobileappdev/homeworkplanner/ClassListFragment.kt
package com.mobileappdev.homeworkplanner

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.activity.addCallback
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.R
import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

class ClassListFragment : Fragment() {

    private var mCrimeRecyclerView: RecyclerView? = null
    private var mAdapter: ClassAdapter? = null
    private var mAddClassButton: Button? = null
    private var mShowAssignmentsButton: Button? = null
    private var classScheduleInitialized = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_class_list, container, false)

        mCrimeRecyclerView = view.findViewById<View>(R.id.class_recycler_view) as RecyclerView
        mCrimeRecyclerView!!.layoutManager = LinearLayoutManager(activity)
        mAddClassButton = view.findViewById<View>(R.id.add_class_button) as Button
        mAddClassButton!!.setOnClickListener {
            val intent = Intent(activity, AddClassActivity::class.java)
            startActivity(intent)
        }

        mShowAssignmentsButton = view.findViewById(R.id.show_assignment_button)
        mShowAssignmentsButton!!.setOnClickListener {
            val intent = Intent(activity, AssignmentListActivity::class.java)
            startActivity(intent)
        }

        mAdapter = ClassAdapter(ClassSchedule.mClasses)
        mCrimeRecyclerView!!.adapter = mAdapter

        val callback = requireActivity().onBackPressedDispatcher.addCallback(this) {
            FirebaseAuth.getInstance().signOut()
            startActivity(Intent(activity, LoginUserActivity::class.java))
            activity!!.finish()
        }

        callback.isEnabled = true

        return view
    }

    override fun onResume() {
        super.onResume()
        GlobalScope.launch {
            coroutineScope {
                updateUI()
            }
        }
    }

    private inner class ClassHolder(inflater: LayoutInflater, parent: ViewGroup) : RecyclerView.ViewHolder(inflater.inflate(R.layout.list_item_class, parent, false)), View.OnClickListener {
        private val mTitleTextView: TextView
        private var mClass: Class? = null


        override fun onClick(view: View) {
            val classInfoAct = Intent(activity, ClassInfoActivity::class.java)
            val bundle = Bundle().apply {
                putString("className", mClass!!.className)
                putString("startTime", mClass!!.startTime)
                putString("endTime", mClass!!.endTime)
                putLong("creditHours", mClass!!.creditHours)
                putStringArrayList("classDays", mClass!!.daysOfWeek)
            }
            classInfoAct.putExtra("classInfo", bundle)
            startActivity(classInfoAct)
        }

        init {
            itemView.setOnClickListener(this)

            mTitleTextView = itemView.findViewById<View>(R.id.class_title) as TextView
        }

        fun bind(aClass: Class) {
            mClass = aClass
            mTitleTextView.text = mClass!!.className
        }
    }

    private fun updateUI() = runBlocking {
        if (!classScheduleInitialized) {
            val lol = GlobalScope.launch { ClassSchedule.initRoutine() }
            lol.join()
            classScheduleInitialized = true
        }
        activity!!.runOnUiThread {
            mCrimeRecyclerView!!.adapter!!.notifyDataSetChanged()
        }
    }

    private inner class ClassAdapter(private val mClasses: List<Class>) : RecyclerView.Adapter<ClassHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ClassHolder {
            val layoutInflater = LayoutInflater.from(activity)

            return ClassHolder(layoutInflater, parent)
        }

        override fun onBindViewHolder(holder: ClassHolder, position: Int) {
            val aClass = mClasses[position]
            holder.bind(aClass)
        }

        override fun getItemCount(): Int {
            return mClasses.size
        }
    }
}
